src/btree/bt_sync.c

```diff
@@ -100,6 +100,48 @@ __sync_dup_walk(WT_SESSION_IMPL *session, WT_REF *walk, uint32_t flags, WT_REF *
     return (0);
 }
 
+/*
+ * __sync_ref_is_obsolete --
+ *     Return whether the ref is obsolete according to the newest stop time pair.
+ */
+static bool
+__sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    WT_ADDR *addr;
+
+    addr = ref->addr;
+
+    return (
+      addr != NULL && __wt_txn_visible_all(session, addr->newest_stop_txn, addr->newest_stop_ts));
+}
+
+/*
+ * __sync_ref_mark_deleted --
+ *     Lock the ref and change state as deleted. Also mark the parent page as dirty.
+ */
+static int
+__sync_ref_mark_deleted(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    /* Ignore root pages as they are never be deleted */
+    if (__wt_ref_is_root(ref))
+        return (0);
+
+    /* Guard against marking a page to be deleted when its reconciliation state is not empty. */
+    if (ref->page && ref->page->modify && ref->page->modify->rec_result != WT_PM_REC_EMPTY)
+        return (0);
```

Comment:
Both `ref.page` and `ref.page.modify` are pointers, and so should be compared to NULL, not treated like booleans.

Reply:
Fixed.